### PR TITLE
netdev_ieee802154: Use intermediate layer for mac header

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR splits the construction of the ieee802154 MAC header. `gnrc_netif_ieee802154` constructs a generic header with only the MAC addresses. The netdev_ieee802154 code then reformats it to a "proper" ieee802.15.4 MAC frame with the information from the device driver.

### Issues/PRs references

#9330 and part of the follow up work of #8198 

WIP as the gomach and the lwmac still need to be adapted.

### CODE REVIEW

**Comment by bergzand** [comment]
Effective impact in code size is around 184 bytes with gnrc_networking on a samr21-xpro

**Comment by miri64** [comment]
For a better architecture, I think *only* ~200 bytes are a good sacrafice.

**Comment by miri64** [comment]
But maybe put a `data` instead of `mac` somewhere in the name. Because implementing MAC with this layer is next to impossible (no way to send beacons, ACKs, etc.)

**Comment by bergzand** [comment]
> But maybe put a data instead of mac somewhere in the name. Because implementing MAC with this layer is next to impossible (no way to send beacons, ACKs, etc.)

Thanks, I was looking for a better name, but couldn't think of anything while implementing.

**Comment by bergzand** [comment]
Part of the L2 statistics are still handled in the gnrc_netif_ieee802154 code.

File: drivers/include/net/netdev/ieee802154.h [hunk 1]
```diff
@@ -121,6 +122,50 @@ typedef struct {
...
+/**
+ * @brief   Data structure for communicating IEEE 802.15.4 send and receive
+ *          call data to upper network layers.
+ */
+typedef struct {
+    uint8_t src[IEEE802154_LONG_ADDRESS_LEN]; /**< source address */
+    uint8_t dst[IEEE802154_LONG_ADDRESS_LEN]; /**< Destination address */
+    uint8_t src_len;                          /**< Source address length */
+    uint8_t dst_len;                          /**< Destination address length */
+} netdev_ieee802154_data_t;
```
Comment (miri64) [inline drivers/include/net/netdev/ieee802154.h]: `data_hdr` instead?

File: sys/net/gnrc/netif/gnrc_netif_ieee802154.c [hunk 2]
```diff
@@ -166,16 +154,11 @@ static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif)
...
-    netdev_ieee802154_t *state = (netdev_ieee802154_t *)netif->dev;
     gnrc_netif_hdr_t *netif_hdr;
-    const uint8_t *src, *dst = NULL;
     int res = 0;
-    size_t src_len, dst_len;
-    uint8_t mhr[IEEE802154_MAX_HDR_LEN];
-    uint8_t flags = (uint8_t)(state->flags & NETDEV_IEEE802154_SEND_MASK);
-    le_uint16_t dev_pan = byteorder_btols(byteorder_htons(state->pan));
+    const uint8_t *dst, *src;
+    netdev_ieee802154_data_t l2data;
```
Comment (miri64) [inline sys/net/gnrc/netif/gnrc_netif_ieee802154.c]: `l2data_hdr`?

File: drivers/include/net/netdev/ieee802154.h (hunk 1 above)
Comment (bergzand) [reply drivers/include/net/netdev/ieee802154.h]: Ok!

File: sys/net/gnrc/netif/gnrc_netif_ieee802154.c (hunk 2 above)
Comment (bergzand) [reply sys/net/gnrc/netif/gnrc_netif_ieee802154.c]: And ok!

**Comment by bergzand** [comment]
@miri64 And another set of renaming commits :)

File: drivers/include/net/netdev/ieee802154.h [hunk 3]
```diff
@@ -121,6 +122,50 @@ typedef struct {
  */
 typedef struct netdev_radio_rx_info netdev_ieee802154_rx_info_t;
 
+/**
+ * @brief   Data structure for communicating IEEE 802.15.4 send and receive
+ *          call data to upper network layers.
+ */
+typedef struct {
+    uint8_t src[IEEE802154_LONG_ADDRESS_LEN]; /**< source address */
```
Comment (miri64) [inline drivers/include/net/netdev/ieee802154.h]: Consistency of capitalization with comments below?

File: drivers/netdev_ieee802154/netdev_ieee802154.c [hunk 4]
```diff
@@ -50,6 +51,65 @@ static int _get_iid(netdev_ieee802154_t *dev, eui64_t *value, size_t max_len)
...
+    if ((res = ieee802154_set_frame_hdr(mhr, data->src, data->src_len,
+                                        data->dst, data->dst_len,
+                                        dev_pan, dev_pan,
+                                        flags, netdev->seq++)) == 0) {
+        DEBUG("netdev_ieee802154: Error preperaring frame\n");
+        return -EINVAL;
+    }
+    /* prepare iolist for netdev / mac layer */
+    iolist_t iolist = {
+        .iol_next = (iolist_t *)list->iol_next,
```
Comment (miri64) [inline drivers/netdev_ieee802154/netdev_ieee802154.c]: This cast shouldn't be necessary.

File: drivers/include/net/netdev/ieee802154.h [hunk 5]
```diff
@@ -121,6 +122,50 @@ typedef struct {
...
+/**
+ * @brief   Receive function for netdev IEEE 802.15.4 devices.
+ *
+ * Reformats the ieee802.15.4 mac frame to a generic
+ * @ref netdev_ieee802154_mac_t header using the frame content.
+ *
+ * @param[in]   dev     network device descriptor
+ * @param[out]  buf     buffer to write into or NULL
+ * @param[in]   len     maximum number of bytes to read
+ * @param[out] info     status information for the received packet. Might
```
Comment (miri64) [inline drivers/include/net/netdev/ieee802154.h]: Okay, since you decided for left aligned below, this one needs adaption as well ;-)

Comment (bergzand) [reply drivers/include/net/netdev/ieee802154.h]: Done! :)

**Comment by bergzand** [comment]
@miri64 I've adapted both lwmac and gomach to also use the new netdev layer.

**Comment by smlng** [comment]
@bergzand so no WIP anymore, and thus ready for review? If so, please change title ...

**Comment by bergzand** [comment]
> @bergzand so no WIP anymore, and thus ready for review? If so, please change title ...

Yup, ready for review, removed the tag from the title

**Comment by miri64** [comment]
> Tested for both "vanilla" GNRC and with both MAC protocols (using the gnrc_networking_mac example). Vanilla works, MAC doesn't (can't ping other node).

(I also can confirm that it *is* working in master).

**Comment by bergzand** [comment]
@miri64 Thank you for your review.

My aim for this PR was to have a it as a preparation step for the layered approach. I first want to have gnrc_netif and netdev decoupled, as in, no direct access to netdev data by gnrc_netif. As a next step in a follow up I want to make the netdev_ieee802154 and netdev_eth a "proper" netdev layer with their own data structs containing a netdev_driver.
As the netdev layer doesn't have a netdev_driver (yet), I can't use it here, thus the hardcoded calls to `netdev_ieee802154_send/recv`

> Tested for both "vanilla" GNRC and with both MAC protocols (using the gnrc_networking_mac example). Vanilla works, MAC doesn't (can't ping other node).

Whoops :disappointed:

**Comment by miri64** [comment]
> As the netdev layer doesn't have a netdev_driver (yet), I can't use it here, thus the hardcoded calls to `netdev_ieee802154_send/recv`

Why not introduce it here. `netdev_eth` should be dealt with in a separate PR anyway.

**Comment by tcschmidt** [comment]
@bergzand IMO it would be beneficial to show a clear description of your attempted software architecture (layering) incl. a drawing. As is it appears slightly unspecified.

**Comment by bergzand** [comment]
> @bergzand IMO it would be beneficial to show a clear description of your attempted software architecture (layering) incl. a drawing. As is it appears slightly unspecified.

Do the schematics from #8198 answer your request? Mostly the diagram and accompanying posts [here](https://github.com/RIOT-OS/RIOT/pull/8198#issuecomment-350449232).

This PRs and any PR related to this one such as #9330 slowly work to the architecture described there.

**Comment by tcschmidt** [comment]
Well, #8198 does not display an architecture. There is some rational in issue #7736 (should be referenced) why netdev needs layering., but little reasoning about the specifics.  It could be envisioned that others who might pick this up later will find it difficult to understand what and why.

**Comment by bergzand** [comment]
> Why not introduce it here. netdev_eth should be dealt with in a separate PR anyway.

Yes, possible. This would change the scope to not only this `netdev_ieee802154_data_hdr_t`, but would also require changes to the current set of 802154 radios.

An implementation of a `netdev_driver_t` struct for the netdev_ieee802154 layer requires a layered structure not only for the send/recv functions, but also for the get/set methods. At the moment these functions are layered in a reverse way (first the function from the radio is called, then the netdev generic function). This would have to be modified to match the send/recv functions (first netdev generic, then the radio). 

I have to see if I also require #9329 for this or that I have to statically allocate the new driver structs in auto_init.

I'll open a new PR with the full changes for this in the next few days.

**Comment by bergzand** [comment]
> Well, #8198 does not display an architecture. There is some rational in issue #7736 (should be referenced) why netdev needs layering., but little reasoning about the specifics. It could be envisioned that others who might pick this up later will find it difficult to understand what and why.

I'm not sure I understand what kind of information you're looking for here. Is your question about the changes in this PR specific or more about the overall architecture that we're aiming at in the long run? If it is the last case, maybe we can move this discussion to #7736.

Are you looking for a graphical explanation on how the different networking layers work together? So how the `gnrc_netif` code calls the netdev layers and how the individual netdev layers interact?

**Comment by tcschmidt** [comment]
@bergzand it's not about me, it is the demand for a crystal clear layout of what is concepted here so that current work and future contributions can be measured against. From the start, netdev is an interface and the goal of making it multilayered sounds weird. 
Presenting a clear, well justified architecture

>   fundamental concepts or properties of a system in its environment embodied in its elements, relationships, and in the principles of its design and evolution (ISO/IEC/IEEE 42010)

may help - and netdev++ was lacking this anyway.

**Comment by tcschmidt** [comment]
@bergzand could we do a thorough practical evaluation of the performance impacts prior to progressing (e.g., like in https://arxiv.org/abs/1801.02833) ? I'm saying this after @PeterKietzmann and I have spent life-month of disclosing and debugging insufficiencies of previous versions of netdev ...

**Comment by bergzand** [comment]
@tcschmidt I really prefer to continue this discussion in #7736 as your questions appear to be more aimed at the overal idea of the layered approach to netdev. This PR is only a small part of the migration.

**Comment by tcschmidt** [comment]
@bergzand fine with me - even though my concern is related to code: We've learned that nice neaties in the code can completely mess up performance ...

**Comment by bergzand** [comment]
@miri64 I was thinking about how to remove the `memmove` from the receive path, the impact on the latency is not really negligible (17us for large frames).
My current best solution is to add a `netdev_data_hdr_t::offset` to indicate where the next layer data (6lowpan header) starts. In the receive path this would be set to `ieee802154_get_frame_hdr_len(pdu_start)` and in the send path it can be zero. The whole `sizeof(header)+header->offset` snip can then be removed by the `gnrc_netif_ieee802154::recv` code.

**Comment by miri64** [comment]
I think, except for the a major API change (where recv would get a list as a parameter) this would be the cleanest approach for the moment...

**Comment by miri64** [comment]
(but maybe call it `mhr_len` instead of offset ;-)).

**Comment by bergzand** [comment]
> (but maybe call it mhr_len instead of offset ;-)).

:+1:

**Comment by miri64** [comment]
Marked as WIP, since you seem to steer this into a different direction according to https://github.com/RIOT-OS/RIOT/issues/7736#issuecomment-438273328.

**Pull request closed without merging**